lightningd/plugin_hook.c

```diff
@@ -89,6 +103,42 @@ void plugin_hook_unregister_all(struct plugin *plugin)
 
 /* Mutual recursion */
 static void plugin_hook_call_next(struct plugin_hook_request *ph_req);
+static void plugin_hook_callback(const char *buffer, const jsmntok_t *toks,
+				 const jsmntok_t *idtok,
+				 struct plugin_hook_request *r);
+
+/* We get notified if a plugin was killed while it was part of a call
+ * chain. If it was still to be called we just remove it from the list,
+ * otherwise it was the plugin that was currently handling the hook call, and
+ * we need to fail over to the next plugin.
+*/
+static void plugin_hook_killed(struct plugin_hook_call_link *link)
+{
+	struct plugin_hook_call_link *head;
+
+	head = list_top(&link->req->call_chain, struct plugin_hook_call_link,
+			list);
+
+	/* If we are the head of the call chain, then the plugin died while it
+	 * was handling the hook call. Pretend it didn't get the memo by
+	 * calling the next one instead. This is correct since it is
+	 * equivalent to the plugin dying before the hook invokation, assuming
+	 * the plugin has not commmitted any changes internally. This is the
+	 * weakest assumption we can make short of restarting the plugin and
+	 * calling the hook again (potentially crashing the plugin the same
+	 * way again.
+	 */
+	if (link == head) {
+		/* Call next will unlink, so we don't need to. This is treated
+		 * equivalent to the plugin returning a continue-result.
+		 */
+		plugin_hook_callback(NULL, NULL, NULL, link->req);
```

Comment:
We get called due to `plugin_killed` being called or because we `tal_free(plugin)` which itself prints a warning to the logs and cleans up / unregisters the plugin from all hooks, hook-calls, and other things they might have registered for.